http-clients/aws-crt-client/src/main/java/software/amazon/awssdk/http/crt/internal/response/CrtResponseAdapter.java

```diff
@@ -39,97 +40,110 @@ public final class CrtResponseAdapter implements HttpStreamResponseHandler {
 
     private final HttpClientConnection connection;
-    private final CompletableFuture<Void> responseComplete;
-    private final AsyncExecuteRequest sdkRequest;
-    private final SdkHttpResponse.Builder respBuilder = SdkHttpResponse.builder();
-    private final int windowSize;
-    private CrtResponseBodyPublisher respBodyPublisher;
+    private final CompletableFuture<Void> completionFuture;
+    private final SdkAsyncHttpResponseHandler responseHandler;
+    private final SimplePublisher<ByteBuffer> responsePublisher = new SimplePublisher<>();
 
-    private CrtResponseAdapter(HttpClientConnection connection,
-                               CompletableFuture<Void> responseComplete,
-                               AsyncExecuteRequest sdkRequest,
-                               int windowSize) {
-        this.connection = Validate.notNull(connection, "HttpConnection is null");
-        this.responseComplete = Validate.notNull(responseComplete, "reqComplete Future is null");
-        this.sdkRequest = Validate.notNull(sdkRequest, "AsyncExecuteRequest Future is null");
-        this.windowSize = Validate.isPositive(windowSize, "windowSize is <= 0");
-    }
+    private final SdkHttpResponse.Builder responseBuilder = SdkHttpResponse.builder();
 
-    public static HttpStreamResponseHandler toCrtResponseHandler(HttpClientConnection connection,
-                                                                 CompletableFuture<Void> responseComplete,
-                                                                 CrtRequestContext request) {
-        return new CrtResponseAdapter(connection, responseComplete, request.sdkRequest(), request.readBufferSize());
+    private CrtResponseAdapter(HttpClientConnection connection,
+                               CompletableFuture<Void> completionFuture,
+                               SdkAsyncHttpResponseHandler responseHandler) {
+        this.connection = Validate.paramNotNull(connection, "connection");
+        this.completionFuture = Validate.paramNotNull(completionFuture, "completionFuture");
+        this.responseHandler = Validate.paramNotNull(responseHandler, "responseHandler");
     }
 
-    private void initRespBodyPublisherIfNeeded(HttpStream stream) {
-        if (respBodyPublisher == null) {
-            respBodyPublisher = new CrtResponseBodyPublisher(connection, stream, responseComplete, windowSize);
-        }
+    public static HttpStreamResponseHandler toCrtResponseHandler(HttpClientConnection crtConn,
+                                                                 CompletableFuture<Void> requestFuture,
+                                                                 SdkAsyncHttpResponseHandler responseHandler) {
+        return new CrtResponseAdapter(crtConn, requestFuture, responseHandler);
     }
 
     @Override
-    public void onResponseHeaders(HttpStream stream, int responseStatusCode, int blockType, HttpHeader[] nextHeaders) {
-        initRespBodyPublisherIfNeeded(stream);
```

Comment:
I didn't try to understand why it was written the way it was.

Thinking about it now:
* Theoretically, we could lazily create the publisher.
* That may make the failure path faster, but I expect that it would slow down the successful path (now all reads of the publisher need to be volatile). 
* It adds complexity.
* It may not even make the failure path faster, because initializing a lazy variable without guarantees about the calling thread is two volatile reads, a lock acquisition, and a volatile write. I'd have to test to be sure, but I expect that's slower than any initialization logic in the publisher's constructor.

I don't think it's worth the amount of time spent testing to see if it's faster, since it's more complex and I doubt that it's any significant speed improvement. The next time we do an optimization path, we can optimize it at that time if it shows up as something significant on the profiler.